### ISSUE TITLE
lint flat_map into filter_map

### ISSUE THREAD

**Issue by xfix** [rust-lang/rust-clippy#2241]
`filter_map` is a more specific version of `flat_map`. It would be neat if Clippy recommended using `filter_map` instead of `flat_map` when the function in `flat_map` returns `Option`.

This is an example from documentation with `filter_map` replaced by `flat_map`.

```rust
fn main() {
    let a = ["1", "2", "lol"];

    let mut iter = a.iter().flat_map(|s| s.parse().ok());

    assert_eq!(iter.next(), Some(1));
    assert_eq!(iter.next(), Some(2));
    assert_eq!(iter.next(), None);
}
```

it would be more readable with `filter_map` instead of `flat_map`.

**Comment by killercup** [comment]
Eh, quick question: Why not just use `a.iter().flat_map(|s| s.parse());`? `Result::iter` only yield the `Ok` value, just like `Option::iter` only yields a value when it's `Some`.

(And here comes the usual E-hard suggestion: Write a lint that detects unnecessary method calls, like the `.ok()` here.)

**Comment by clarcharr** [comment]
IMO now that `Try` is stable, upstream should replace `filter_map` with a version that uses `Try` instead. That would allow use of `Result` as well as option.

**Comment by Manishearth** [comment]
You can't change the signatures of trait methods upstream

**Comment by clarcharr** [comment]
...right, that's correct. :/

**Comment by oli-obk** [comment]
> You can't change the signatures of trait methods upstream

you can if it's generic and strictly allows more code

**Comment by Manishearth** [comment]
I'm talking about trait methods. You can do that with inherent methods.

This is a trait you can implement yourself. Most don't override flat_map,
but they *can*, so we can't change this.

On Dec 2, 2017 12:46 PM, "Oliver Schneider" <notifications@github.com>
wrote:

You can't change the signatures of trait methods upstream

you can if it's generic and strictly allows more code

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/rust-lang-nursery/rust-clippy/issues/2241#issuecomment-348718949>,
or mute the thread

**Comment by clarcharr** [comment]
I mean, they *could* always do a crater run to check. Although at the moment, isn't `FilterMap` literally coded as a façade over `FlatMap`, or am I wrong?

**Comment by Manishearth** [comment]
No, that's not one of the "allowed" kinds of breakages.

Crater is not a carte blanche to make breaking changes.

There's a set of allowed breaking changes. We still do crater runs for
those.

Additionally for fixing bugs and soundless issues we are allowed to make
otherwise-disallowed breakages with a crater run, but this is not a biggie.

**Comment by xfix** [comment]
`FilterMap` has a better implementation of `size_hint` than `FlatMap` as it knows that `next` can only provide 0 or 1 element and doesn't store currently processed iterator in it as it doesn't need to do that with `Option`.

**Comment by llogiq** [comment]
One thing we could lint is manually constructing one-element-`vec![..]`s in `flat_map`, where using `filter_map` will do. Example: https://github.com/rust-lang/rust/pull/55476
